crates/sui-adapter/src/adapter.rs

```diff
@@ -1100,6 +1106,134 @@ pub fn resolve_and_type_check(
+                ExecutionErrorKind::entry_argument_error(idx, EntryArgumentErrorKind::TypeMismatch),
+                format!(
+                "Function expects {}::{}::{} struct but provided argument's value does not match",
+                string_struct.0, string_struct.1, string_struct.2,
+            ),
+            )
+        })?;
+    validate_string_move_value(&string_move_value, idx, string_struct.1)
+}
+
+// Given a MoveValue representing an string argument (string itself or arbitrarily nested vector of
+// strings), validate that the content of the data represents a valid string
+fn validate_string_move_value(
+    value: &MoveValue,
+    idx: LocalIndex,
+    module: &IdentStr,
+) -> Result<(), ExecutionError> {
+    match value {
+        MoveValue::Vector(vec) => {
+            for v in vec {
+                validate_string_move_value(v, idx, module)?;
+            }
+            Ok(())
+        }
+        MoveValue::Struct(MoveStruct::Runtime(vec)) => {
+            // deserialization process validates the structure of this MoveValue (one struct field
+            // in string structs containing a vector of u8 values)
+            debug_assert!(vec.len() == 1);
+            if let MoveValue::Vector(u8_vec) = &vec[0] {
+                validate_string(&move_values_to_u8(u8_vec), idx, module)
+            } else {
+                debug_assert!(false);
+                Ok(())
+            }
+        }
+        _ => {
+            debug_assert!(false);
+            Ok(())
+        }
+    }
+}
+
+// Converts a Vec<MoveValue::U8> to a Vec<U8>
+fn move_values_to_u8(values: &[MoveValue]) -> Vec<u8> {
+    let res: Vec<u8> = values
+        .iter()
+        .filter_map(|v| {
+            // deserialization process validates the structure of this MoveValue (u8 stored in a
+            // vector of a string struct)
+            if let MoveValue::U8(b) = v {
+                Some(b)
+            } else {
+                None
+            }
+        })
+        .cloned()
+        .collect();
+    debug_assert!(res.len() == values.len());
+    res
+}
+
+// Validates that Vec<u8> represents a valid string
+fn validate_string(bytes: &[u8], idx: LocalIndex, module: &IdentStr) -> Result<(), ExecutionError> {
+    if module == STD_ASCII_MODULE_NAME {
+        for b in bytes {
+            if *b > 0x7F {
+                return Err(ExecutionError::new_with_source(
+                    ExecutionErrorKind::entry_argument_error(
+                        idx,
+                        EntryArgumentErrorKind::TypeMismatch,
+                    ),
+                    format!("Unexpected non-ASCII value in argument {}", idx),
+                ));
+            }
+        }
+    } else {
+        debug_assert!(module == STD_UTF8_MODULE_NAME);
+        if str::from_utf8(bytes).is_err() {
+            return Err(ExecutionError::new_with_source(
+                ExecutionErrorKind::entry_argument_error(idx, EntryArgumentErrorKind::TypeMismatch),
```

Comment:
I think it's fine for now, but we might want to give some more information here? 

It's also probably pretty rare people will hit this unless they are creating tooling, so maybe it doesn't matter

Reply:
Yes, this should be pretty rare, but I added a bit more info in the error messages.